nexus/reconfigurator/planning/src/planner.rs

```diff
@@ -1681,6 +1758,125 @@ impl<'a> Planner<'a> {
+
+        // If we don't actually have a TUF repo here, we can't do
+        // updates anyway; any return value is fine.
+        if new_repo.tuf_repo().is_none() {
+            return Ok(report);
+        }
+
+        // Check that all in-service zones (other than Nexus) on all
+        // sleds have an image source consistent with `new_repo`.
+        for sled_id in self.blueprint.sled_ids_with_zones() {
+            for z in self.blueprint.current_sled_zones(
+                sled_id,
+                BlueprintZoneDisposition::is_in_service,
+            ) {
+                let kind = z.zone_type.kind();
+                if kind != ZoneKind::Nexus
+                    && z.image_source != new_repo.zone_image_source(kind)?
+                {
+                    report.set_waiting_on(
+                        NexusGenerationBumpWaitingOn::NonNexusZoneUpdate,
+                    );
+                    return Ok(report);
+                }
+            }
+        }
+
+        // Confirm that we have new nexuses at the desired generation number
+        let current_generation = self.blueprint.nexus_generation();
+        let proposed_generation = self.blueprint.nexus_generation().next();
+        let mut out_of_date_nexuses_at_current_gen = 0;
+        let mut nexuses_at_next_gen = 0;
+        for sled_id in self.blueprint.sled_ids_with_zones() {
+            for z in self.blueprint.current_sled_zones(
+                sled_id,
+                BlueprintZoneDisposition::is_in_service,
+            ) {
+                if let BlueprintZoneType::Nexus(nexus_zone) = &z.zone_type {
+                    if nexus_zone.nexus_generation == proposed_generation {
+                        nexuses_at_next_gen += 1;
+                    }
+
+                    if nexus_zone.nexus_generation == current_generation
+                        && z.image_source
+                            != new_repo.zone_image_source(z.zone_type.kind())?
+                    {
+                        out_of_date_nexuses_at_current_gen += 1;
+                    }
+                }
+            }
+        }
+
+        if out_of_date_nexuses_at_current_gen == 0 {
+            // If all the current-generation Nexuses are "up-to-date", then we may have
+            // just completed handoff successfully. In this case, there's nothing to report.
+            return Ok(report);
+        } else {
+            // If there aren't enough Nexuses at the next generation, quiescing could
+            // be a dangerous operation. Blueprint execution should be able to continue
+            // even if the new Nexuses haven't started, but to be conservative, we'll wait
+            // for the target count.
+            if nexuses_at_next_gen < self.input.target_nexus_zone_count() {
+                report.set_waiting_on(
+                    NexusGenerationBumpWaitingOn::NewNexusBringup,
+                );
+                return Ok(report);
```

Comment:
I actually do not have test coverage for this case, and would like to add it before this PR merges. I have struggled to do it through zone manipulation - because Nexus is discretionary, we'll be eager to add the new Nexus zones if we can (and why not? They should wait on boot for handoff).

To force this to happen, I'm thinking I'll need to construct a scenario where we expunge a sled so that we **cannot** actually place this new Nexus, and observe that the handoff does not occur while we're operating at a reduced capacity.